### DISCUSSION TITLE
Is it possible to record the screen from any operating system?

### DISCUSSION THREAD

**Discussion by juanibonini** [doronz88/pymobiledevice3#385] [Q&A]
Hi,

I need to record a screen from any operative system and I would like to know, if it is posible to do it in a similar way to how a screenshot is taken.

Thanks in advance!!

**Comment by rorlig** [comment]
I have same use case, wondering if this is possible. ?

**Comment by doronz88** [comment]
Are you referring to screen video or a single image?

**Reply by rorlig** [reply]
Screen video recording not a single image

**Reply by doronz88** [reply]
In that case, just as I said in #765:
I know only of https://github.com/danielpaulus/quicktime_video_hack
But never managed to get it work

**Reply by rorlig** [reply]
me neither :(

**Comment by nanoscopic** [comment]
There are a variety of ways to "record the screen".

1. Screenshots through lockdown. People have monkeyed with this to create a mjpeg stream but it is slow because the screenshots are large and need to be re-encoded to jpeg. It's unpleasant.

2. The Valeria protocol usb interface. This is what Quicktime Video Hack uses. Daniel RE'ed this for Sauce Labs and made a Java implementation for them. Later he released the Go version. The implementation he released works but crashes often in my experience. I couldn't depend on it. I recently wrote my own implementation of Valeria protocol... on recent MacOS systems they lock down access to the interface so you can't access it any more though, at least not without some sort of USB interposer.

3. XCUITest. Running xctest's have access to an API that gives screenshots in a way better than through lockdown. Careful coding can lead to a smoother stream of images that way than through lockdown screenshots.

4. The official "screen recording extensions". You can make an app that records the screen. It has to be activated, but this is the official method and has the best results so far that I've seen.

5. Xcode "Debug View Hierarchy" can fetch images of all the elements of an app that can be lldb attached. These can be combined to show a screenshot. In theory this could be used to make a new type of screencast method.

6. Whatever "iPhone Mirroring" is doing. I haven't looked into it just yet.

7. Airplay / Screen casting. This is basically the same as 4, but more wonky. There are open source implementations but they involve hacky implementations of Apple's encryption nonsense. I would not recommend this.

9. Other variants of the Valeria protocol, such as used by carplay. I've poked at this a bit... but not figured much out yet.
